### PR TITLE
ARROW-2963: [C++] Make thread pool fork-safe

### PR DESCRIPTION
**Pull request by pitrou**
In a forked child process, ThreadPool should reinitialize its internal state as though no threads had been spawned.

### CODE REVIEW

File: cpp/src/arrow/util/thread-pool.cc [hunk 1]
```diff
@@ -51,15 +51,46 @@ struct ThreadPool::State {
...
+#endif
+}
 
 ThreadPool::~ThreadPool() {
   if (shutdown_on_destroy_) {
     ARROW_UNUSED(Shutdown(false /* wait */));
   }
 }
 
+void ThreadPool::FixIfBroken() {
```
Comment (cpcloud) [inline cpp/src/arrow/util/thread-pool.cc]: Is there a less vague name for what this does?

File: cpp/src/arrow/util/thread-pool.h [hunk 2]
```diff
@@ -136,6 +140,8 @@ class ARROW_EXPORT ThreadPool {
   void LaunchWorkersUnlocked(int threads);
   // Get the current actual capacity
   int GetActualCapacity();
+  // Protect against fork()
```
Comment (cpcloud) [inline cpp/src/arrow/util/thread-pool.h]: Like maybe `ProtectAgainstFork()`? :)

File: cpp/src/arrow/util/thread-pool.cc [hunk 3]
```diff
@@ -186,6 +220,7 @@ void ThreadPool::WorkerLoop(std::shared_ptr<State> state,
 
 Status ThreadPool::SpawnReal(std::function<void()> task) {
   {
+    FixIfBroken();
```
Comment (wesm) [inline cpp/src/arrow/util/thread-pool.cc]: What is the rough cost of this (e.g. # of nanoseconds)? I guess this can be measured in the benchmarks

Comment (wesm) [reply cpp/src/arrow/util/thread-pool.cc]: According to https://stackoverflow.com/a/41783404 it is < 10ns. This seems OK but we should definitely keep an eye on benchmarks for thread spawning overhead

**Comment by wesm** [comment]
+1 modulo the potential for renaming the Fork Protector, and assuming there aren't concerns about performance

**Comment by cpcloud** [comment]
> +1 modulo the potential for renaming the Fork Protector, and assuming there aren't concerns about performance

Protecting all things tine-y.

I think we need a meme with Antoine on it.

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**